OutOfSchool/OutOfSchool.WebApi/Services/Elasticsearch/ElasticsearchSynchronizationService.cs

```diff
@@ -0,0 +1,183 @@
+            {
+                logger.LogError($"Syncronization of Elasticsearch has failed during {ElasticsearchSyncOperation.Delete} operation");
+                return false;
+            }
+
+            logger.LogInformation($"Synchronization of Elasticsearch has finished successfully.");
+
+            return true;
+        }
+
+        public async Task Create(ElasticsearchSyncRecordDto dto)
+        {
+            var elasticsearchSyncRecord = mapper.Map<ElasticsearchSyncRecord>(dto);
+            try
+            {
+                await elasticsearchSyncRecordRepository.Create(elasticsearchSyncRecord).ConfigureAwait(false);
+
+                logger.LogInformation("ElasticsearchSyncRecord created successfully.");
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                logger.LogError($"Creating new record to ElasticserchSyncRecord failed.");
+                throw;
+            }
+        }
+
+        public async Task AddNewRecordToElasticsearchSynchronizationTable(ElasticsearchSyncEntity entity, Guid id, ElasticsearchSyncOperation operation)
+        {
+            ElasticsearchSyncRecordDto elasticsearchSyncRecordDto = new ElasticsearchSyncRecordDto()
+            {
+                Entity = entity,
+                RecordId = id,
+                OperationDate = DateTimeOffset.UtcNow,
+                Operation = operation,
+            };
+
+            await Create(elasticsearchSyncRecordDto).ConfigureAwait(false);
+        }
+
+        public async Task Synchronize(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                logger.LogInformation("Elasticsearch synchronization started");
+
+                await Synchronize().ConfigureAwait(false);
+
+                logger.LogInformation("Elasticsearch synchronization finished");
+
+                await Task.Delay(10000, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<bool> Synchronize(IEnumerable<ElasticsearchSyncRecord> elasticsearchSyncRecords, ElasticsearchSyncOperation elasticsearchSyncOperation)
+        {
+            var ids = elasticsearchSyncRecords.Where(es => es.Operation == elasticsearchSyncOperation).Select(es => es.RecordId);
+
+            if (!ids.Any())
+            {
+                return true;
+            }
+
+            if (elasticsearchSyncOperation == ElasticsearchSyncOperation.Delete)
+            {
+                var result = await esProvider.DeleteRangeOfEntitiesByIdsAsync(ids).ConfigureAwait(false);
+
+                if (result == Result.Error)
+                {
+                    logger.LogError($"Error happend while trying to delete indexes in Elasticsearch.");
+                    return false;
+                }
+            }
+            else
+            {
+                var workshops = await databaseService.GetByIds(ids).ConfigureAwait(false);
```

Comment:
so we're getting workshop entities, convert them to dtos and then convert them to ES models. We are not using the DTOs in any way except to convert to ES. maybe a dedicated mapper Entity -> ES model will be a better solution?